controllers/managedclusterForCGU_controller.go

```diff
@@ -0,0 +1,207 @@
+/*
+Copyright 2021.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	clusterv1 "github.com/open-cluster-management/api/cluster/v1"
+	policiesv1 "github.com/open-cluster-management/governance-policy-propagator/api/v1"
+	ranv1alpha1 "github.com/openshift-kni/cluster-group-upgrades-operator/api/v1alpha1"
+)
+
+const (
+	clusterNotReadyRetryDelay = time.Minute * 5
+	clusterNoStatusRetryDelay = time.Minute * 1
+	ztpInstallNS              = "ztp-install"
+	ztpDeployWaveAnnotation   = "ran.openshift.io/ztp-deploy-wave"
+)
+
+// ManagedClusterReconciler reconciles a ManagedCluster object to auto create the ClusterGroupUpgrade
+type ManagedClusterForCguReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+// Reconcile the managed cluster auto create ClusterGroupUpgrade
+// - When a new managed cluster is created, create ClusterGroupUpgrade CR for the cluster only when it's ready
+//   and its child policies are available
+//
+// Note: The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ManagedClusterForCguReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
```

Comment:
It would be helpful to document the conditions which cause this method to get called.